### DISCUSSION TITLE
ST7789 Wrong Colors on Display

### DISCUSSION THREAD

**Discussion by webdjoe** [arendst/Tasmota#21908] [Q&A] [answered]
I am using an ST7789 240x320 SPI color display with a ESP32-S3. I've tried all of the ST7789 display descriptor files (including the esp32-s2 example) but the display becomes distorted and has missing chunks with all of the ST7789 examples. I managed to get the display working to a point by editing the `0:`, `:1`, `:2` and `:3` opcodes to:

```
:0,C0,00,00,00
:1,A0,00,00,01
:2,00,00,00,02
:3,60,00,00,03
```

But now there is a color issue where the background color is light yellow when it is supposed to be white. When using `displaytext`, the text shows up as light cyan.  I tried to use displaytext with red and it shows as light yellow. I changed the `:i,21,20` line to `:i,20,21` and now the colors are completely inverted (black <> white, background and text). Much easier to see but still wonky when setting colors. Does anyone have any ideas on what I can try? 

Does anyone know of any other resources or documentation on the display opcodes?

**Comment by sfromis** [comment, marked as answer]
This thread also includes making a 240x240 ST7789 display work with Tasmota: https://github.com/arendst/Tasmota/discussions/21791

What I did was to use the default [ST7789 display descriptor](https://github.com/arendst/Tasmota/blob/development/tasmota/displaydesc/ST7789_display.ini) (stripping off the touch part), and then `DisplayInvert 1` to flip the inverted colors. And jkwim mentioned fixing it in the display descriptor, changing `:i,20,21` to `:i,21,20`.

**Reply by webdjoe** [reply]
Thank you for the reply, I think I have it! The [ST7789_display.ini](https://github.com/arendst/Tasmota/blob/development/tasmota/displaydesc/ST7789_display.ini) descriptor without the touch interface did not work right either. The screen was distorted and had static rectangles running through the screen. I used the [esp32s3-esp32-1732S019_display.ini](https://github.com/arendst/Tasmota/blob/development/tasmota/displaydesc/esp32s3-esp32-1732S019_display.ini) as a base, which gave me the least bit of distortion, and then I edited the `:0, :1, :2, :3` sections until there was no more distortion. I have no idea what I was actually doing but was able to make it work by trial and error. I'm not sure if it has to do with using an esp-s3 or this ST7789 has a different display size from the example. 

When `displayinvert 0`,  the `:i,21,20` has the same result as not having a `:i` line, the boot background color is black on splash and startup. The text color is correct when using `DisplayText [zf1s2x80y80Ci2]Red[x80y112Ci3]Green[x80y144Ci4]Blue`. The pages.jsonl background and text colors are also correct.  This is my end result:

Here is the final display.ini:
```
:H,ST7789,240,320,16,SPI,1,*,*,*,*,*,*,*,40
:S,2,1,3,0,80,30
:I
01,A0
11,A0
3A,81,55
36,81,00
21,80
13,80
29,A0
:o,28
:O,29
:A,2A,2B,2C
:R,36
:0,C0,00,00,00
:1,A0,00,00,01
:2,00,00,00,02
:3,60,00,00,03
:r,1
:B,20,4
#
```

Using `DisplayText [zf1s2x80y80Ci2]Red[x80y112Ci3]Green[x80y144Ci4]Blue`
![unnamed (Small)](https://github.com/user-attachments/assets/6a224f1c-58fd-4ce2-a568-2cd40d7b332a)

With rudimentary and poorly designed haspmota `pages.jsonl` with no sensor values feeding:
```
{"page":0,"comment":"---------- Upper stat line ----------"}

{"id":11,"obj":"label","x":0,"y":0,"w%":100,"pad_right":100,"h":22,"bg_color":"#000033", "text_color": "#FFFFFF", "bg_opa":255,"radius":0,"border_side":0,"text":"Bedroom","text_font":"montserrat-10"}

{"id":16,"obj":"lv_clock","x%":60,"y":3,"w%":30,"h":16,"radius":0,"border_side":0, "text_color": "#FFFFFF"}

{"page":1,"comment":"---------- Page 1 ----------"}
{"id":0,"bg_color":"#FFFFFF","text_color":"#000000"}

{"id":2,"obj":"arc","x":10,"y":20,"w":65,"h":40,"border_side":0,"type":0,"rotation":0,"start_angle":180,"end_angle":0,"start_angle1":180,"value_font":12,"value_ofs_x":0,"value_ofs_y":-14,"bg_opa":0,"text":"--.-°F","min":32,"max":100,"val":0,"val_rule":"ESP32#Temperature","val_rule_formula":"val * 10","text_rule":"ESP32#Temperature","text_rule_format":"%2.1f °F"}

{"id":10,"obj":"label","x":172,"y":35,"w":140,"text":"MPU","align":0}
{"id":11,"obj":"label","x":172,"y":55,"w":140,"text":"x=","align":0,"text_rule":"tele-BME820#humidity","text_rule_format":"x=%6.3f"}
{"id":12,"obj":"label","x":172,"y":75,"w":140,"text":"y=","align":0,"text_rule":"tele-BME820#pressure","text_rule_format":"y=%6.3f"}
```

![IMG_0474 (Small)](https://github.com/user-attachments/assets/9e65a8d9-f4a2-48a7-9ecc-79d96d7cc583)

**Reply by sfromis** [reply]
When it comes to making the sensor updates work, you need to follow the Berry format for rule triggers (instead of "classic" rules), as HASPmota is created using this language.
https://tasmota.github.io/docs/Berry/#teleperiod-rules

However, unless you really only want updates at `TelePeriod`, I'd recommend to leave out the `tele#` prefix, giving you quick updates of the displayed values.
